### PR TITLE
Use DynamoDB for DCDO & Gatekeeper in test web application server

### PR DESCRIPTION
**Pull request by sureshc**
The transition from the unicorn web application server to puma in 2018 broke the logic that decides what storage system to use for the DCDO and Gatekeeper dynamic configuration systems. Fix a hard coded reference to `unicorn`.

Previously, our intention was to use the in memory adapter for DCDO and Gatekeeper when running unit tests on development environments, continuous integration builds, and the managed test server (test.code.org / test-studio.code.org), to use DynamoDB for production (code.org / studio.code.org), and to use the file system adapter when running the web application server on the managed test server. If/when we switch the managed test server to use multiple, autoscaled, load balanced web application servers, the file system store won't work because each web application server could potentially have a different value for a given setting.

New rule:

- production always uses DynamoDB
- managed test server (test.code.org / test-studio.code.org) uses DynamoDB when executing the puma web application server
- managed test server uses Memory when running unit tests
- development environment uses Memory when running unit tests
- continuous integration environment uses Memory when running unit tests and when executing the web application server (note that continuous integration builds [use puma to run the web application](https://github.com/code-dot-org/code-dot-org/blob/41c593d6e51553d067cf90e24796945283de036c/lib/rake/circle.rake#L93) and could be switched to use the File System adapter in the future, if needed).

This change supports the upgrade from Rails 5.2 to Rails 6.0 because without being able to set the Gatekeeper setting that enables splitting read databases, we can't run browser integration tests with the new Rails 6 read splitting functionality enabled.

This change also supports moving the projects tables from Pegasus to Dashboard because [the transition logic](#44957 ) relies on DCDO settings to change the name of the tables that our queries use.
## Testing story

Variations of this change have been tested manually on the managed test server by directly modifying the code.

## Deployment strategy

I’ve backed up the production DCDO and Gatekeeper DynamoDB Tables by exporting them as CSVs.
There are a number of keys that control critical features or foundational settings. Corruption or accidental replacement these settings due to a flaw in this Pull Request’s logic could significantly disrupt our production systems. I’ll prep a Hotfix Pull Request.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: lib/dynamic_config/dcdo.rb [hunk 1]
```diff
@@ -69,9 +69,11 @@ def self.create
     env = Rails.env.to_s if defined?(Rails) && Rails.respond_to?(:env)
 
     cache_expiration = 5
-    if env == 'test'
+    # Use the memory adapter if we're running unit tests, but not if we're running the web application server.
+    if env == 'test' && File.basename($0) != 'puma'
       adapter = MemoryAdapter.new
-    elsif env == 'production'
+    # Production and the managed test system web application servers (test.code.org / studio.code.org) use DynamoDB.
+    elsif env == 'production' || (CDO.test_system? && File.basename($0) == 'puma')
```
Comment (sureshc) [inline lib/dynamic_config/dcdo.rb]: Note that `CDO.test_system?` does not support `staging-next`. Should we update that method to return true for staging-next?

**Review by Hamms — COMMENTED** [review]
The core idea here sounds good to me, but can the repeated logic here get DRYed out into some shared helper methods?

**Comment by cat5inthecradle** [comment]
> The core idea here sounds good to me, but can the repeated logic here get DRYed out into some shared helper methods?

I'm torn on this because chasing down exactly what each helper means can be tedious too (like, what is `test_system`). That's might all be resolved by proper IDE setup. Normally I'd completely agree, but I'm thinking ahead to the kinds of troubleshooting typically done in these files, and what would be the most useful helpers. What kinds of method names were you thinking?

**Comment by Hamms** [comment]
Here's my proposed change: https://github.com/code-dot-org/code-dot-org/pull/45256

**Review by Hamms — APPROVED** [review]
This LGTM as a fix for the existing issue.

As follow-up work, I'd suggest we also add unit tests for DCDOBase and GatekeeperBase. Also, it looks like these two classes share quite a bit of functionality which could be split out, either into shared concerns or a base class.

**Comment by Hamms** [comment]
Merged staging latest to resolve a merge conflict introduced by https://github.com/code-dot-org/code-dot-org/pull/45418/files#diff-9d6502fdbc36cecfe029344c72579bfa407731327c21d8f5da94dee14289788bR5

**Pull request merged**